galsim/bandpass.py

```diff
@@ -83,85 +83,116 @@ class Bandpass(object):
         # `wave_list` is normally constructed (see `Bandpass.__mul__` below)
 
-        # Figure out input throughput type.
-        tp = throughput  # For brevity within this function
-        if isinstance(tp, basestring):
-            import os
-            if os.path.isfile(tp):
-                tp = galsim.LookupTable(file=tp, interpolant='linear')
-            else:
-                # Evaluate the function somewhere to make sure it is valid before continuing on.
-                if red_limit is not None:
-                    test_wave = red_limit
-                elif blue_limit is not None:
-                    test_wave = blue_limit
-                else:
-                    # If neither `blue_limit` nor `red_limit` is defined, then the Bandpass should
-                    # be able to be evaluated at any wavelength, so check.
-                    test_wave = 700
-                try:
-                    tp = eval('lambda wave : ' + tp)
-                    tp(test_wave)
-                except:
-                    raise ValueError(
-                        "String throughput must either be a valid filename or something that "+
-                        "can eval to a function of wave. Input provided: {0}".format(throughput))
+        self._orig_tp = throughput  # Save this for pickling.
+        self._tp = _tp              # This is orig_tp turned into an actual function
+
+        if blue_limit > red_limit:
+            raise ValueError("blue_limit must be less than red_limit")
+        self.blue_limit = blue_limit # These may change as we go through this.
+        self.red_limit = red_limit
+        self.zeropoint = zeropoint
+        self.wave_type = wave_type
 
         # Figure out wavelength type
-        if wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
-            wave_factor = 1.0
-        elif wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
-            wave_factor = 10.0
+        if self.wave_type.lower() in ['nm', 'nanometer', 'nanometers']:
+            self.wave_factor = 1.0
+        elif self.wave_type.lower() in ['a', 'ang', 'angstrom', 'angstroms']:
+            self.wave_factor = 10.0
         else:
-            raise ValueError("Unknown wave_type '{0}'".format(wave_type))
+            raise ValueError("Unknown wave_type '{0}'".format(self.wave_type))
+
+        # Convert string input into a real function (possibly a LookupTable)
+        self._initialize_tp()
+
+        if _wave_list is not None:
+            # Manual override!  Be careful!
+            self.wave_list = _wave_list
+            # This also means that red_limit and blue_limit are already set correctly.
+            # Don't change them.
+            assert self.blue_limit is not None
+            assert self.red_limit is not None
+            return
+
+        # Account for wave_factor in wavelength limits
+        if self.wave_factor != 1.9:
```

Comment:
I think this should be `1.0` not `1.9`.  Though I guess this makes the if statement unnecessary since if `self.wave_factor` *is* 1.0, then dividing by it is harmless.